### DISCUSSION TITLE
Invalid extrinsics are stored in the block

### DISCUSSION THREAD

**Discussion by Mr-Leshiy** [paritytech/substrate#9651] [Q&A] [answered]
According to this wiki https://wiki.polkadot.network/docs/maintain-errors, and how extrinsics are processed in the substrate. I have found that they even the extrinsic become invalid it is still added into the blockchain, but without the state mutating. 
What was an intention to store an invalid extrinsics inside blocks ? Does  it lead to the redundant storage usage of the unneeded and unuseful data ?
Are you considering to change this and store inside block only valid extrinsics ?

**Comment by gui1117** [comment]
> I have found that they even the extrinsic become invalid it is still added into the blockchain, but without the state mutating.

what do you mean precisely by invalid extrinsic ? a failing extrinsic (i.e. an extrinsic which returns a dispatch error) is still included in the block and usually does contains some state mutation like the payment of the fee of the extrinsic for instance.

**Reply by Mr-Leshiy** [reply]
Yes, I mean this extrinsic with returns dispatch error.

**Reply by KiChjang** [reply]
Could you be more specific as to which invalid extrinsics that you saw that didn't mutate any state?

**Comment by shawntabrizi** [comment, marked as answer]
You add "invalid" extrinsics into the block because you still need to charge a fee.

This is the same reason you add transaction which "run out of gas" on the Ethereum network.

If the blockchain would not include the extrinsic or charge a fee, then it becomes easy to attack the chain with bad transactions which still does computation and waste resources of the node.

**Reply by Mr-Leshiy** [reply]
As I see, actually currently it is also possible to make such attack with the transactions with the invalid signature, so actually I suppose that you already have solution even for such scenario. Of course I understood that inside the extrinsic it can be possible much more heavier computations in comparison with the just signature validation.
Maybe better to do it in the more configurable way and give an opportunity for the developer to choose, which scenario he will choose, do not add invalid extrinsic inside the block and do not take any fee (like it is done in the Bitcoin) or include this extrinsic charge fee.

**Reply by xlc** [reply]
developers can already optin to disallow inclusion of bad transactions by implementing a custom signed extension.

**Reply by Mr-Leshiy** [reply]
@xlc , ohh it seems I have missed this one, thank you for the answer, it is definitely what I am looking for. Can you pls provide maybe some docs or links to code how it is possible to do.

**Reply by xlc** [reply]
I don't know if there are good docs about SignedExtension but here is a usage: https://github.com/paritytech/substrate/blob/152f76459e57e8ed2d6d921825c2b08d35294651/frame/transaction-payment/src/lib.rs#L600

**Reply by h4x3rotab** [reply]
@Mr-Leshiy I think we are the only team using SignedExtension for this purpose:

https://github.com/Phala-Network/phala-blockchain/blob/master/pallets/phala/src/mq/check_seq.rs
